### PR TITLE
OCF: remove presence, add support for Links and Collections, add polling algorithms

### PR DESCRIPTION
**Pull request by zolkis**
in favor of a future solution, and for observing /oic/res for resource deletion. Fixing #92.

Signed-off-by: Zoltan Kis <zoltan.kis@intel.com>

### CODE REVIEW

**Comment by zolkis** [comment]
@gabrielschulhof , @poussa PTAL

File: api/ocf/client.md [hunk 1]
```diff
@@ -72,14 +87,17 @@ Note that applications should not create `ClientResource` objects, as they are c
...
 
 <a name="onresourcelost"></a>
-The `delete` event is fired on a `ClientResource` object when the implementation gets notified about the resource being deleted or unregistered from the OCF network. This might not be supported in all OCF networks.
+The `delete` event is fired on a `ClientResource` object when the implementation gets notified about the resource being deleted or unregistered from the OCF network. The event listener receives a [`ResourceId`](#resourceid) dictionary object that contains the `deviceId` and `resourcePath` of the deleted resource.
+
+###### Note
+Note that presence (`/oic/ad` resource and `oic.wk.ad` resource type) is not any more defined by the OCF specification. Therefore implementations SHOULD observe the `/oic/res` resource for getting notified when a resource is removed, acccording to the [`findResources` steps](#findresources). When a resource is deleted, implementations SHOULD fire the `delete` event on the deleted resource.
```
Comment (gabrielschulhof) [inline api/ocf/client.md]: -not any more
+no longer

File: api/ocf/README.md [hunk 2]
```diff
@@ -93,12 +93,17 @@ Exposes information about the OCF platform that hosts the current device.
...
 
-OCF errors (see also [these notes](../README.md#errors)) are represented as augmented [`Error`](https://nodejs.org/api/errors.html#errors_class_error) objects with added properties. The following [`Error` names](https://nodejs.org/api/errors.html) are used for signaling OCF issues:
-- `OcfDiscoveryError`
-- `OcfObserveError`
-- `OcfPresenceError`.
+OCF errors (see also [these notes](../README.md#errors)) are represented as augmented instances of [`Error`](https://nodejs.org/api/errors.html#errors_class_error) objects.
 
-All these errors are instances of [`Error`](https://nodejs.org/api/errors.html#errors_class_error) and contain the following additional properties:
+The following [`Error` names](https://nodejs.org/api/errors.html) are used for signaling OCF issues:
+- `OcfResourceNotFound` used if the resource if cannot be located in the OCF network, or when an observed resource is deleted from the network.
```
Comment (nagineni) [inline api/ocf/README.md]: >  used **if** the resource **if** cannot...

remove one if

**Comment by nagineni** [comment]
Since the presence is gone, is there any reason to keep `devicelost` event in client?

**Comment by zolkis** [comment]
> Since the presence is gone, is there any reason to keep devicelost event in client?

We can remove it, and re-add later when the device entry/exit proposal is accepted.

Or, we could keep it there but it won't work until it's fixed in the spec. Of course implementation could keep working with `/oic/ad` but it will be adherent to an earlier version of the OIC Core spec.

**Comment by gabrielschulhof** [comment]
Github now allows us to push to a PR if we are collaborators. Cool!

**Comment by zolkis** [comment]
What github didn't know is that I have amended my last commit _after_ you pushed yours :D

**Comment by gabrielschulhof** [comment]
@zolkis indeed! Feel free to delete my commit.

**Comment by zolkis** [comment]
@gabrielschulhof could we merge this?

**Comment by gabrielschulhof** [comment]
@zolkis Please specify that the `polling` value is to be given in milliseconds.

**Comment by zolkis** [comment]
> Please specify that the polling value is to be given in milliseconds.

It is already there: client.md line 91.

**Comment by gabrielschulhof** [comment]
@zolkis oh, there it is :)

**Comment by gabrielschulhof** [comment]
LGTM 👍

**Comment by gabrielschulhof** [comment]
Actually, I take back the "LGTM" for now by deleting the comment where I gave it, because `polling` should maybe not be specified per-resource. It makes no sense to issue multiple recurring requests to `/oic/res` on a given device for different resources on that device at potentially different intervals.

Instead, it makes sense to have a single recurring GET /oic/res listener per device, started whenever the first `"delete"` listener is attached to any resource on that device and removed whenever the last `"delete"` listener is removed from the resources on said device.

**Comment by zolkis** [comment]
> It makes no sense to issue multiple recurring requests to /oic/res on a given device for different resources on that device at potentially different intervals.

I agree with this, but this has little to do with the [polling algorithm](https://github.com/01org/iot-js-api/blob/7ea4004e2e831cbbb95c7eb997b209e58372607b/api/ocf/client.md#resourcepolling). That is about  a different polling than the one replacing presence (polling `/oic/res`), the former is a possible step in the [latter](https://github.com/01org/iot-js-api/blob/7ea4004e2e831cbbb95c7eb997b209e58372607b/api/ocf/client.md#resourcestatus) and is triggered when polling `/oic/res` does not work (for any reason).

Whether a resource is polled or not (for value, or for testing existence), and with what interval, should be specified per resource, because it is resource specific (how often the resource is expected to change, or how often the client app wants to know about change).

When `polling` is not specified, it is up to the implementation whether and how to use polling to respect the contracts of this API.

When `polling` is specified by the application, it is a (resource dependent) hint to the implementation that can be taken into account or be overridden by the implementation.

Does that make sense?

**Comment by gabrielschulhof** [comment]
Wait a sec - what now? I thought the only reason for implementing polling was so we poll `/oic/res` once per device for the purposes of implementing the `"delete"` event.

**Comment by gabrielschulhof** [comment]
Oh, I see ... if polling `/oic/res` does not work, poll the individual resource. Right?

**Comment by zolkis** [comment]
> Oh, I see ... if polling /oic/res does not work, poll the individual resource. Right?

Yes! And the implementation decides on that, depending on the capabilities of the underlying stack.

**Comment by gabrielschulhof** [comment]
Well, in this implementation it's very unlikely that polling /oic/res won't
work. Observing /oic/res will certainly not work, but polling /oic/res will
almost certainly work.

**Comment by zolkis** [comment]
> polling /oic/res will almost certainly work

No problem, then you won't need to implement resource polling for presence.

The decision on polling `/oic/res` and polling individual resources are kind of independent, since the latter can be also triggered by applications (as a different feature) when they set `resource.polling` - but the implementation can override that too, depending on implementation (system) context.
Too flexible/confusing?

**Comment by gabrielschulhof** [comment]
I would say so. If a resource does not support observation, we shouldn't provide polling for it. Let the applications implement that themselves.

Polling for presence is a necessity. Polling for data is not.

**Comment by zolkis** [comment]
Of course presence (and polling, either for presence or value) makes sense only for resources that are under observation...

We have 2 separate needs to provide app-settable polling interval:
- presence via polling either `oic/res`, or the resource itself;
- polling the resource anyway (presumably with larger interval) regardless whether observe is on or off (most likely on). 

One might argue this second use case is not needed. But even if the implementation can be trusted to provide the `ondelete` event when the resource is unreachable, the impl still needs to know what is the interval after which a resource is declared dead. That is resource dependent and app-context dependent, so optimally the app should have a way of telling that to the impl.

Keeping `resource.polling` would also allow implementations to choose the best polling times per remote device.

If we remove `resource.polling`, then polling becomes entirely encapsulated by implementations (or how should we allow apps to provide an interval hint for that), and as a result, for the second use case apps will poll themselves, which IMHO is not desirable.

We could leave this open for longer consideration if needed.

**Comment by gabrielschulhof** [comment]
OK, so, like, "The resource hasn't given me any new data in > resource.polling, so it must be dead" ... ?

**Comment by zolkis** [comment]
No, it's more like "the resource has not given me any new data in `resource.polling` time, so let's check", and if that retrieve fails, the resource is dead. The implementation should do this, in order to be able to fire the `resource.ondelete` event, but it would be nice if the app would have some effect on the value of `resource.polling`.

**Comment by gabrielschulhof** [comment]
OK, gotcha. But what if the retrieve doesn't fail? Do I convert the response to an `"update"` event? Do I assume that the properties have the same value as the last update event and can therefore discard the response?

Also, I would prefer that this checking be opt-in rather than opt-out. That is, I would not perform this check unless `resource.polling` were explicitly set to a positive value, because I want the default to be not to perform this check, rather than to always perform this check at, say, 5 second intervals *unless* `resource.polling` is set to 0.

**Comment by zolkis** [comment]
It is opt-in.

By default, `resource.polling` is `undefined`. The implementation decides whether to do polling or not in order to implement `resource.ondelete`.

If at some point the app sets `resource.polling` to 0, the impl stops polling the resource itself altogether and relies only on polling `/oic/res`.

If at some point the app sets `resource.polling` to a positive value, the impl starts polling. If retrieve fails, the impl fires the `resource.ondelete` and unwinds the resource (stops observing etc).
If retrieve succeeds, we have two options:
1. the impl checks the resource representation; if changed, fires the `resource.onupdate` event (like saying "I know the resource is alive, but since no change happened, no events are fired".
2. the impl fires the `resource.onupdate` event anyway, like saying "here is your data you asked at least every `resource.polling` intervals".
I am somewhat open to which semantic we implement, but option 1 would be closer to the original contract of `resource.onupdate`. If the app would do polling itself, would do the resource repr comparison itself. So I say we should go by option 1.

**Comment by gabrielschulhof** [comment]
Actually, come to think of it, I wouldn't fire `onupdate` even if the data had changed, because, presumably the resource will fire notify in such an event, and the `onupdate` will be fired "naturally".

**Comment by zolkis** [comment]
Right, so it would be enough to do the first part of the steps (including the clause for "if retrieve fails, ...".

**Comment by zolkis** [comment]
Based on the discussion above, it seems the polling part is more or less stable, so merging this in order to have presence removed and errors defined. Polling algorithms can be refined in later commits.

**Pull request merged**